File: internal/models/lot_purchase.go
package models

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

type LotPurchase struct {
	ID                   uint           `gorm:"primaryKey" json:"id"`
	SellerName           string         `gorm:"column:seller_name" json:"seller_name"`
	PurchaseDate         time.Time      `gorm:"column:purchase_date" json:"purchase_date"`
	Description          string         `gorm:"type:text" json:"description"`
	TotalCost            float64        `gorm:"column:total_cost" json:"total_cost"`
	EstimatedMarketValue float64        `gorm:"column:estimated_market_value" json:"estimated_market_value"`
	LotContentSnapshot   string         `gorm:"type:text" json:"lot_content_snapshot"`
	Status               string         `json:"status"`
	TrackedItems         []TrackedItem  `gorm:"foreignKey:LotPurchaseID" json:"tracked_items"`
	CreatedAt            time.Time      `json:"created_at"`
	UpdatedAt            time.Time      `json:"updated_at"`
	DeletedAt            gorm.DeletedAt `gorm:"index" json:"-"`
}

func (l *LotPurchase) ParseSnapshot() ([]SnapshotItem, error) {
	if l.LotContentSnapshot == "" {
		return nil, nil
	}
	var items []SnapshotItem
	if err := json.Unmarshal([]byte(l.LotContentSnapshot), &items); err != nil {
		return nil, err
	}
	return items, nil
}

File: src/main/java/com/collectingwithzak/entity/LotPurchase.java
package com.collectingwithzak.entity;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.collectingwithzak.dto.SnapshotItem;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "lot_purchases")
@Getter
@Setter
@SQLDelete(sql = "UPDATE lot_purchases SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
public class LotPurchase extends BaseEntity {

    @Column(name = "seller_name")
    private String sellerName;

    private String description;

    @Column(name = "purchase_date")
    private LocalDate purchaseDate;

    @Column(name = "total_cost", columnDefinition = "numeric(10,2)")
    private double totalCost;

    @Column(name = "estimated_market_value", columnDefinition = "numeric(10,2)")
    private double estimatedMarketValue;

    @Column(name = "lot_content_snapshot", columnDefinition = "text")
    private String lotContentSnapshot;

    private String status = "PENDING";

    @OneToMany(mappedBy = "lotPurchase", fetch = FetchType.LAZY)
    private List<TrackedItem> trackedItems = new ArrayList<>();

    public List<SnapshotItem> parseSnapshot() {
        if (lotContentSnapshot == null || lotContentSnapshot.isBlank()) {
            return List.of();
        }
        try {
            ObjectMapper mapper = new ObjectMapper();
            return mapper.readValue(lotContentSnapshot, new TypeReference<>() {});
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse lot snapshot", e);
        }
    }
}
